### PR TITLE
XLNet Modules

### PR DESCRIPTION
**Pull request by AvinashBukkittu**
This PR adds the following

- XLNet Encoder along with test cases

- XLNet Classifier built using XLNet Encoder along with test cases

- XLNet Regressor built using XLNet Encoder along with test cases

- All XLNet and BERT related modules are packaged in `pretrained` folder

### CODE REVIEW

File: texar/modules/classifiers/xlnet_classifier.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+            if self._hparams.use_projection:
+                self.projection = layers.get_layer(hparams={
+                    "type": "Dense",
+                    "kwargs": {
+                        "units": self._encoder.output_size
+                    }
+                })
+
+            # summary type
+            if self._hparams.summary_type == 'last':
```
Comment (ZhitingHu) [inline texar/modules/classifiers/xlnet_classifier.py]: This part need to be updated (see the pytorch version) -- `summary_type` --> `cls_strategy`

File: texar/modules/classifiers/xlnet_classifier.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+# pylint: disable=too-many-arguments, invalid-name, no-member,
+# pylint: disable=too-many-branches, too-many-locals, too-many-statements
+
+__all__ = [
+    "XLNetClassifier"
+]
+
+
+class XLNetClassifier(ClassifierBase):
```
Comment (ZhitingHu) [inline texar/modules/classifiers/xlnet_classifier.py]: The pytorch version has a `param_group` method https://github.com/asyml/texar-pytorch/blob/master/texar/modules/classifiers/xlnet_classifier.py#L173

What's the counterpart here?

File: texar/modules/classifiers/xlnet_classifier.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            "num_classes": 2,
+            "logit_layer_kwargs": None,
+            "summary_type": "last",
+            "dropout": 0.1,
+            "use_projection": True,
+            "name": "xlnet_classifier"
+        })
+        return hparams
+
+    def _build(self, token_ids, segment_ids=None, input_mask=None, mode=None):
```
Comment (ZhitingHu) [inline texar/modules/classifiers/xlnet_classifier.py]: add docstring.  `_build` will be shown in the doc (e.g., https://texar.readthedocs.io/en/latest/code/modules.html#texar.modules.Conv1DClassifier._build)

File: texar/modules/classifiers/xlnet_classifier.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        if mode is None:
+            mode = global_mode()
+        output, _ = self._encoder(token_ids, segment_ids, input_mask=input_mask,
+                                  mode=mode)
+        summary = self.summary_op(output)
+        if self._hparams.use_projection:
+            summary = tf.tanh(self.projection(summary))
+        # summary: (batch_size, hidden_dim)
+        summary = self._dropout_layer(
+            summary, training=(mode == tf.estimator.ModeKeys.TRAIN))
```
Comment (ZhitingHu) [inline texar/modules/classifiers/xlnet_classifier.py]: `mode` should be a tensor. Here is a good reference of how `mode` is used: https://github.com/asyml/texar/blob/master/texar/modules/embedders/embedders.py#L191

File: texar/modules/encoders/xlnet_encoders.py [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+        attn_mask_pad = tf.zeros([qlen, mlen], dtype=dtype)
+        ret = tf.concat([attn_mask_pad, mask_u - mask_dia], axis=1)
+        if same_length:
+            mask_l = tf.matrix_band_part(attn_mask, -1, 0)
+            ret = tf.concat([ret[:, :qlen] + mask_l - mask_dia, ret[:, qlen:]],
+                            axis=1)
+
+        return ret
+
+    def _build(self, inputs, *args, **kwargs):
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: `_build` is exposed to users while `_execute` is not. Pls move `_execute`'s docstring to `_build`. Users are not aware of the existence of `_execute`. Take the pytorch version as a reference.

File: texar/modules/pretrained/xlnet_model_utils.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    def _post_attention(self, attn_vec):
+        shape = tf.shape(attn_vec)
+        attn_vec = tf.reshape(
+            tensor=attn_vec,
+            shape=[shape[0], shape[1], self.num_heads*self.head_dim])
+        attn_out = self.output_projection(attn_vec)
+        attn_out = self.dropout(attn_out)
+        return attn_out
+
+    def _build(self, states_h, pos_embed, states_g=None, segment_mat=None,
```
Comment (ZhitingHu) [inline texar/modules/pretrained/xlnet_model_utils.py]: Add docstring to all `_build` methods, on par w/ `forward` in pytorch

Comment (ZhitingHu) [reply texar/modules/pretrained/xlnet_model_utils.py]: There should be an explicit `mode` argument instead of `**kwargs` which is harder to read.

File: texar/modules/regressors/regressor_base.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        return {
+            "name": "classifier"
+        }
+
+    def _build(self, inputs, *args, **kwargs):
+        """Runs regressors on inputs.
+
+        Args:
+          inputs: Inputs to the classifier.
```
Comment (ZhitingHu) [inline texar/modules/regressors/regressor_base.py]: "classifier" --> "regressor"

File: texar/modules/regressors/xlnet_regressor.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            if self._hparams.use_projection:
+                self.projection = layers.get_layer(hparams={
+                    "type": "Dense",
+                    "kwargs": {
+                        "units": self._encoder.output_size
+                    }
+                })
+
+            # summary type
+            if self._hparams.summary_type == 'last':
```
Comment (ZhitingHu) [inline texar/modules/regressors/xlnet_regressor.py]: Same comments as in `XLNetClassifier`

File: texar/modules/regressors/xlnet_regressor.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        hparams.update({
+            "logit_layer_kwargs": None,
+            "summary_type": "last",
+            "dropout": 0.1,
+            "use_projection": True,
+            "name": "xlnet_regressor"
+        })
+        return hparams
+
+    def _build(self, token_ids, segment_ids=None, input_mask=None, mode=None):
```
Comment (ZhitingHu) [inline texar/modules/regressors/xlnet_regressor.py]: Same comments as in XLNetClassifier

File: docs/code/modules.rst [hunk 10]
```diff
@@ -82,6 +82,12 @@ Encoders
 .. autoclass:: texar.modules.RNNEncoderBase
     :members:
 
+:hidden:`XLNetEncoder`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+.. autoclass:: texar.modules.XLNetEncoder
+    :members:
+    :exclude-members: _execute
```
Comment (ZhitingHu) [inline docs/code/modules.rst]: Do not expose `_execute` to users, only `_build`

File: texar/modules/classifiers/xlnet_classifier.py (hunk 1 above)
Comment (AvinashBukkittu) [reply texar/modules/classifiers/xlnet_classifier.py]: Done! [9280366]

File: texar/modules/classifiers/xlnet_classifier.py (hunk 4 above)
Comment (AvinashBukkittu) [reply texar/modules/classifiers/xlnet_classifier.py]: Done! [9280366]

File: texar/modules/encoders/xlnet_encoders.py (hunk 5 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Done! [9280366]

File: texar/modules/pretrained/xlnet_model_utils.py (hunk 6 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_model_utils.py]: Done! [9280366]

File: texar/modules/regressors/regressor_base.py (hunk 7 above)
Comment (AvinashBukkittu) [reply texar/modules/regressors/regressor_base.py]: Done! [9280366]

File: texar/modules/regressors/xlnet_regressor.py (hunk 8 above)
Comment (AvinashBukkittu) [reply texar/modules/regressors/xlnet_regressor.py]: Done! [9280366]

File: texar/modules/regressors/xlnet_regressor.py (hunk 9 above)
Comment (AvinashBukkittu) [reply texar/modules/regressors/xlnet_regressor.py]: Done! [9280366]

File: texar/modules/classifiers/xlnet_classifier.py (hunk 3 above)
Comment (AvinashBukkittu) [reply texar/modules/classifiers/xlnet_classifier.py]: Done! [3448a9c]

File: docs/code/modules.rst (hunk 10 above)
Comment (AvinashBukkittu) [reply docs/code/modules.rst]: I think exclude members means to not show the function doc

File: texar/modules/encoders/xlnet_encoders.py [hunk 11]
```diff
@@ -0,0 +1,530 @@
...
+                             target_mapping=target_mapping, bi_data=bi_data,
+                             clamp_len=clamp_len, cache_len=cache_len,
+                             same_length=same_length, attn_type=attn_type,
+                             two_stream=two_stream, **kwargs)
+
+    def _execute(self, word_embed, segment_ids=None, input_mask=None,
+                 memory=None, permute_mask=None, target_mapping=None,
+                 bi_data=False, clamp_len=None, cache_len=0, same_length=False,
+                 attn_type='bi', two_stream=False, mode=None):
+        r"""Compute XLNet representations for the input.
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: If the docstring is the same as that of forward, we can simply say this and omit the details. Or, if it's mostly the same, just point out the differences.

FYI, the pytorch PR https://github.com/asyml/texar-pytorch/pull/128 fixed this.

File: texar/modules/pretrained/xlnet_model_utils.py (hunk 6 above)
Comment (ZhitingHu) [reply texar/modules/pretrained/xlnet_model_utils.py]: docstring to all `_build` to be added

File: texar/modules/regressors/regressor_base.py (hunk 7 above)
Comment (ZhitingHu) [reply texar/modules/regressors/regressor_base.py]: hmm.. I still see "classifier" here

Comment (AvinashBukkittu) [reply texar/modules/regressors/regressor_base.py]: Interesting. https://github.com/AvinashBukkittu/texar/blob/xlnet/texar/modules/regressors/regressor_base.py#L41 I can see the change here.

File: texar/modules/encoders/xlnet_encoders.py (hunk 11 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Changes are fixed in f578727

File: texar/modules/pretrained/xlnet_model_utils.py (hunk 6 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_model_utils.py]: Changes are visible in f578727

File: texar/modules/classifiers/xlnet_classifier.py (hunk 2 above)
Comment (AvinashBukkittu) [reply texar/modules/classifiers/xlnet_classifier.py]: I have implemented `param_groups` in a similar way like `texar-pytorch` in f578727. In tensorflow, to apply gradients, we call `optimizer.apply_gradients()` with a tuple containing pairs of `variable` and its corresponding gradient. I think maybe it would be better if `param_groups` returns a dict mapping variables to its learning rate, so that it becomes easier for the user to apply the learning rates to the corresponding gradient values. Let me know what you think. For the purpose of consistency however, we can keep it similar pytorch version.

**Comment by AvinashBukkittu** [comment]
> Have these modules reproduced the XLNet example results?

Yes, I tested it on STS-B and IMDB datasets.

File: texar/__init__.py [hunk 12]
```diff
@@ -23,7 +23,7 @@
 # pylint: disable=wildcard-import
```
Comment (ZhitingHu) [inline texar/__init__.py]: Why is this file needed to be changed at all in this PR?

Comment (ZhitingHu) [reply texar/__init__.py]: This makes conflicts with other PR which is intended to update this file

File: texar/modules/classifiers/xlnet_classifier.py (hunk 2 above)
Comment (ZhitingHu) [reply texar/modules/classifiers/xlnet_classifier.py]: "`param_groups` returns a dict" --- this is better. 

Have we implemented the XLNet example as in PyTorch that demonstrates the use of the modules?

File: texar/modules/encoders/xlnet_encoders.py [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+            lr (float): The learning rate. Can be omitted if
+                :attr:`lr_layer_decay_rate` is 1.0.
+            lr_layer_scale (float): Per-layer LR scaling rate. The `i`-th layer
+                will be scaled by `lr_layer_scale ^ (num_layers - i - 1)`.
+            decay_base_params (bool): If `True`, treat non-layer parameters
+                (e.g. embeddings) as if they're in layer 0. If `False`, these
+                parameters are not scaled.
+
+        Returns:
+            The parameter groups, used as the first argument for optimizers.
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: This statement is too vague. Pls explain more (what's the data structure? what's each field?) and give an example of how to use this APIs.

Comment (ZhitingHu) [reply texar/modules/encoders/xlnet_encoders.py]: If the doc of pytorch version is also this vague, pls update it as well.

File: texar/modules/encoders/xlnet_encoders.py [hunk 14]
```diff
@@ -0,0 +1,602 @@
...
+                    "lr": lr * decay_rate,
+                }
+                param_groups.append(param_group)
+        else:
+            param_groups = self.trainable_variables
+        return param_groups
+
+    @property
+    def output_size(self):
+        r"""Return the output size of the network.
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: This doc should be more specific and clearer

File: texar/modules/encoders/xlnet_encoders.py [hunk 15]
```diff
@@ -0,0 +1,602 @@
...
+        return ret
+
+    def _build(self, token_ids, segment_ids=None, input_mask=None,
+               memory=None, permute_mask=None, target_mapping=None,
+               bi_data=False, clamp_len=None, cache_len=0, same_length=False,
+               attn_type='bi', two_stream=False, mode=None):
+        r"""Compute XLNet representations for the input.
+
+        Args:
+            token_ids: Shape `[batch_size, seq_len]`.
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: In other modules' doc we say `[batch_size, max_time]`, instead of `seq_len`. https://github.com/asyml/texar/blob/master/texar/modules/decoders/rnn_decoders.py#L303

Pls keep the terminology consistent. Also, if PyTorch has the same issue, pls fix too

File: texar/modules/pretrained/xlnet_model_utils.py [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+        # residual + layer norm
+        output = tf.contrib.layers.layer_norm(
+            input + output, begin_norm_axis=-1, scope=self.variable_scope,
+            reuse=tf.AUTO_REUSE)
+
+        return output
+
+
+class PositionalEmbedding(ModuleBase):
+    r"""Sinosoidal Positional Embedding."""
```
Comment (ZhitingHu) [inline texar/modules/pretrained/xlnet_model_utils.py]: This embedding must be moved to `embedder` and if possible, mered with the current Sinosoidal position embedder, pls open an issue to record this TODO

File: texar/modules/pretrained/xlnet_utils.py [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+            initialized_variable_names[model_name + ":0"] = 1
+        else:
+            tf.logging.info('model name:{} not exist'.format(model_name))
+
+    return assignment_map, initialized_variable_names
+
+
+def init_xlnet_checkpoint(init_checkpoint_dir, scope_name):
+    """
+    Initializes XLnet model parameters from a checkpoint.
```
Comment (ZhitingHu) [inline texar/modules/pretrained/xlnet_utils.py]: "Initializes" or "Initialize"? Keep consistent w/ other docstrings

File: texar/__init__.py (hunk 12 above)
Comment (AvinashBukkittu) [reply texar/__init__.py]: I had to change it because the CI for python 2.7 was failing because of weird issues in imports. Here is the link to CI that failed https://travis-ci.org/asyml/texar/jobs/569081378#L1779.

File: texar/modules/classifiers/xlnet_classifier.py (hunk 2 above)
Comment (AvinashBukkittu) [reply texar/modules/classifiers/xlnet_classifier.py]: Okay, I will change `param_groups` to return a dict mapping a `tf.Variable` to its learning rate. I have not implemented the examples, since I needed `XLNetEncoder` to be ready to build the examples.

File: texar/modules/pretrained/xlnet_model_utils.py (hunk 16 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_model_utils.py]: Done. Issue https://github.com/asyml/texar/issues/195

File: texar/modules/pretrained/xlnet_utils.py (hunk 17 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_utils.py]: Done in 056c949

File: texar/modules/encoders/xlnet_encoders.py (hunk 15 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Fixed in 056c949. Issue in texar-pytorch https://github.com/asyml/texar-pytorch/issues/144

File: texar/modules/encoders/xlnet_encoders.py (hunk 14 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Fixed in 8a99af3

File: texar/modules/encoders/xlnet_encoders.py (hunk 13 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Fixed in fbcdc5d. Issue in texar-pytorch https://github.com/asyml/texar-pytorch/issues/145

File: texar/modules/pretrained/xlnet_utils.py (hunk 17 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_utils.py]: Fixed in 8a99af3

File: texar/modules/pretrained/xlnet_model_utils.py (hunk 16 above)
Comment (AvinashBukkittu) [reply texar/modules/pretrained/xlnet_model_utils.py]: Added a TODO. Issue https://github.com/asyml/texar/issues/195

File: texar/modules/encoders/xlnet_encoders.py (hunk 15 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Fixed in 056c949. Issue in texar-pytorch https://github.com/asyml/texar-pytorch/issues/144

File: texar/__init__.py (hunk 12 above)
Comment (AvinashBukkittu) [reply texar/__init__.py]: I just checked the master branch is failing with the same CI error - https://travis-ci.org/asyml/texar/jobs/568599445#L1765

File: texar/modules/encoders/xlnet_encoders.py [hunk 18]
```diff
@@ -0,0 +1,606 @@
...
+        r"""Create parameter groups for optimizers. When
+        :attr:`lr_layer_decay_rate` is not 1.0, parameters from each layer form
+        separate groups with different base learning rates.
+
+        This method should be called before applying gradients to the variables
+        through the optimizer. Particularly, after calling the optimizer's
+        `compute_gradients` method, the user can call this method to get
+        variable-specific learning rates for the network. The gradients for each
+        variables can then be scaled accordingly. These scaled gradients are
+        finally applied by calling optimizer's `apply_gradients` method.
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: The explanation is good but still a bit difficult to understand. Can we add an example piece of code, like [here](https://github.com/asyml/texar/blob/master/texar/modules/decoders/rnn_decoders.py#L144). The example code should be simple

File: texar/modules/encoders/xlnet_encoders.py [hunk 19]
```diff
@@ -0,0 +1,606 @@
...
+            for variable in self.trainable_variables:
+                vars_to_learning_rates[variable] = lr
+
+        return vars_to_learning_rates
+
+    @property
+    def output_size(self):
+        r"""Return the size of the undivided hidden dimension of the network.
+        """
+        return self._hparams.hidden_dim
```
Comment (ZhitingHu) [inline texar/modules/encoders/xlnet_encoders.py]: `output_size` is defined to be "The final dimension(s) of :meth:`forward` output tensor(s)" https://github.com/asyml/texar-pytorch/blob/master/texar/torch/module_base.py#L95

Now the `forward` returns two tensors, why is the output_size here a single size? (Same question applies to the PyTorch version)

File: texar/modules/encoders/xlnet_encoders.py (hunk 18 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Added an example in b667dd6

File: texar/modules/encoders/xlnet_encoders.py (hunk 19 above)
Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: I discussed with Zecong about this. At other places, `output_size` returns an int denoting the size of the network. If we instead return sizes of both tensors of `_build`, the user would have to handle this case differently which may be a bit of a trouble for the user. It is important to note that user actually cares about the size of `output` tensor returned by `_build`. I have added a note to indicate that this method returns the size of the `output` tensor.

Comment (ZhitingHu) [reply texar/modules/encoders/xlnet_encoders.py]: Then the docstring "The final dimension(s) of :meth:forward output tensor(s)" should be updated. It's not exactly the size(s) of forward outputs, but instead more of case-by-case? E.g., sometimes it's the dimensions of some of the output tensors. 

And, what does "final dimension(s)" mean? 

Since it's case-by-case, all `output_size` should explicitly and clearly explain what it is in the particular classes.

Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: I will create an issue in `texar-pytorch` to track and fix this. In `texar-tf`, I just checked, other modules don't actually have `output_size` method. I happened to have copied it from `texar-pytorch`. So we can either keep it or remove it altogether in `texar-tf`.

Comment (ZhitingHu) [reply texar/modules/encoders/xlnet_encoders.py]: Let's keep it here. To make the docstring absolutely clear, we can probably say: 

```
The last dimension of the encoder output.
	
Note: The :meth:`_build` returns two tensors of shapes 
`[batch_size, max_time, hidden_dim]` and `[batch_size, cache_len, hidden_dim]`.
 output_size here equals `hidden_dim`
```

Comment (ZhitingHu) [reply texar/modules/encoders/xlnet_encoders.py]: For class properties (with `@property` decorator), the docstring usually does not say `Return xxx` cuz it's not  used as a function

Comment (AvinashBukkittu) [reply texar/modules/encoders/xlnet_encoders.py]: Fixed in 1ffca93
